### DISCUSSION TITLE
ElementComposite()

### DISCUSSION THREAD

**Discussion by JarosT55** [kinnala/scikit-fem#1087] [Q&A] [answered]
**Hi,**
I would like to use (or experiment in the context of a project similar to ex24) with an element which is composed of **ElementTriP2** and **ElementTriMorley** ... something like:
_eC = ElementComposite(ElementTriP2(), ElementTriMorley(_))
or in its vector form
_eCV = ElementVector(ElementComposite(ElementTriP2(), ElementTriMorle_y()),2)

The current setup with:
_e = ElementVector(ElementTriP2())_
works quite well, but having _'u_n'_ dof type would give me more flexibility and potentially? the better numerical flow.

However, there are some issues I don't know how to address.
Dofs types of an **ElementComposite(...)** are a all-inclusive set of the respective elements' dofs.
They are:
for _eC:   ['u^1', 'u^2', 'u^1', 'u_n^2']_
for _eCV:  ['u^1^1', 'u^1^2', 'u^2^1', 'u^2^2', 'u^1^1', 'u^1^2', 'u_n^2^1', 'u_n^2^2']_
and because the 'all-inclusive' composition of elements resulting in dofs sets contain redundancies, like _u^1 ≡ u^2_
The question.
Is there a way to avoid such redundancies and, e.g. create _eC_ in the form:
_eC:   ['u^1', 'u^1', 'u_n^2']_

**Regards
MichaelT**

**Comment by kinnala** [comment, marked as answer]
That would require programming your own Element class. However, it is not clear to me what kind of finite element (see mathematical definition of a finite element) is this expected to correspond to. Normally you cannot simply drop out DOFs from an element and expect the result to make sense.

**Reply by kinnala** [reply]
There are no "redundacies" as you call them. ElementComposite is used to approximate multiple solutions at once.

**Reply by JarosT55** [reply]
**Hi Mr Kinnala,**

_###Normally   you cannot simply drop out DOFs from an element and expect the result to make   sense.###_   

Yes,   I am aware of this fact. My question was directed towards “automatic or   semiautomatic” generating a n**ew Composite Finite Element Space**, removing   potential redundancies of arguments of the respective fields of composed   elements, in short, a new **ElementClass()**.   In   many cases, it could not make sense, but I suppose adding  **‘u_n’**  to **ElementTriPX()**, would. I have looked at **ElementTriPX()**   classes and others offering **‘u_n’**. They seem to be constructed differently. 
How   difficult would such modification be??? 
Is it feasible? Would it be   reinventing the wheel? 
**I have no clue** … so I have asked the question.        

Thank   you, Mr Kinnala, for the prompt answer.  
     
**Regards  
MichaelT**       

_PS.   I had a short glimpse into **Symfem**  (_a symbolic finite element definition library_) but did not venture there ... yet :)._

**Reply by kinnala** [reply]
I suggest you try to explain what your goal is with 'u' and 'u_n' and then we can think of different solutions. I presume sometimes you want to set 'u' and sometimes 'u_n'? It can be done without special finite element.

**Reply by JarosT55** [reply]
**Hi, Mr Kinnala,**

At this stage, I just want to experiment and better understand the software environment. It was a '**black magic'** weeks ago, now it is a **'very dark grey'**
What is the idea about **'n_u'**? 
I want to check/experiment with how the numerical process responds to setting DOF from **'u'** (which works fine) to **'u_n'**. It would be easier (more elegant) to define **'u_n'** DOF in one element than in the composite one. Obviously, it is also an opportunity to learn the software!!!

I am attaching a sample project's PNG and MP4 (with low-resolution mesh). The goal is to make a transition on the right side of the pink line as 'smooth' as possible, minimizing artefacts (they are smaller for hi-res meshes). It is not really critical; just say ... the esthetic exercise is progressing on the software learning curve. I have tried higher-order **TriPX()**, and the results are positive. 

**Regards
MichaelT**

![GraviTower_E_velocity_HR_0000](https://github.com/kinnala/scikit-fem/assets/23690439/4d7cdf36-e333-401e-a1f5-75583c7091a4)
![GraviTower_E_pressure_HR_0000](https://github.com/kinnala/scikit-fem/assets/23690439/ffefb647-2c1c-4c8a-bbc8-1d99279e4b70)
https://github.com/kinnala/scikit-fem/assets/23690439/f06146cb-134e-4e49-973e-4251dd87821f

**Reply by JarosT55** [reply]
**Hi, Mr Kinnala,**

I have rethought the issue of adding '**u_n'** to **ElementTriPX**'s, and I hope I come to some conclusion on the question I asked, … but by any means, welcome any correction of my line of thought. 

_### How difficult would such modification be???_
_### Is it feasible?_ 
_### Would it be reinventing the wheel?_

The addition of **'u_n'** to **ElementTriPX** where **X**>1 creates a mathematical dilemma of solving polynomial equations of order two or more. Even the quadratic equations possess some challenges with two possible solutions. Hence, we have **TriMorley ≡ TriP1+u_n**, while **TriMorleyX** (for **X**>1) are unavailable.
Constatating the questions … 

_### How difficult would such modification be???_  … **difficult if not … impossible**
_### Is it feasible?_ … **very unlikely**
_### Would it be reinventing the wheel?_ … **no, the wheel in the dimension X>1 does not exist … yet ;).**

… I will put this puzzle to rest :).

**Regards
MichaelT**
